# Spam Detector

In [33]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [34]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
#data.head()
data.shape

(4601, 58)

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*My Hypothesis:*
<p>I hypothesize that the Random Forest Classifier will achieve a higher accuracy score than the Logistic Regression model on the spam detection dataset. This is based on the assumption that the dataset contains complex, non-linear interactions between features, which the Random Forest model is better suited to capture. The accuracy of both models will be evaluated using accuracy_score on a held-out test set."

## Split the Data into Training and Testing Sets

In [35]:
# Create the labels set `y`
y = data['spam']

# and features DataFrame `X`
X = data.drop(columns='spam')


In [36]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_balance = y.value_counts()
print(label_balance)

spam
0    2788
1    1813
Name: count, dtype: int64


In [37]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=11)


## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [38]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [39]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)


In [40]:
# Scale the training data
# Scaling the Test data since the Training data was scaled above
X_test_scaled = scaler.transform(X_test)


## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [41]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression


In [42]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_model = LogisticRegression()
lr_model.fit(X_train_scaled, y_train)
lr_prediction = lr_model.predict(X_test_scaled)

# Review the predictions
lr_prediction[:5]

array([0, 1, 0, 0, 1])

In [43]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
lr_accuracy = accuracy_score(y_test, lr_prediction)
lr_accuracy


0.9296264118158123

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [44]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier


In [45]:
# Make and save testing predictions with the saved logistic regression model using the test data
# Create the model
rf_model = RandomForestClassifier(random_state=11)

# Fit the model to the training data
rf_model.fit(X_train_scaled, y_train)

# Make predictions using the testing data
rf_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_predictions[:5]

array([0, 1, 0, 0, 1])

In [46]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_accuracy = accuracy_score(y_test, rf_predictions)
rf_accuracy

0.9591659426585578

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

**Model Performance Comparison:**
<p>The Random Forest Classifier outperformed the Logistic Regression model, because it achieved a higher accuracy score of 0.9592 compared to the Logistic Regression model's 0.9296.

**Comparison to the Hypothesis:**
<p>The hypothesis predicted that the Random Forest Classifier would perform better because the data might have complex, non-linear interactions between features. The results support this hypothesis, as the Random Forest Classifier performed better.

This evaluation confirms that for this specific dataset, the Random Forest model was more effective at detecting spam compared to the Logistic Regression model.